gdal/frmts/jpegxr/jpegxrdataset.cpp

```diff
@@ -0,0 +1,1202 @@
+        return FALSE;
+
+    if (pabyHeader[3] != 0x00 && pabyHeader[3] != 0x01)
+        return FALSE;
+
+    return TRUE;
+}
+
+/************************************************************************/
+/*                              Open()                                  */
+/************************************************************************/
+
+GDALDataset *JPEGXRDataset::Open( GDALOpenInfo * poOpenInfo)
+{
+    if (!Identify(poOpenInfo) || poOpenInfo->fpL == nullptr)
+        return nullptr;
+
+/* -------------------------------------------------------------------- */
+/*      Create a corresponding GDALDataset.                             */
+/* -------------------------------------------------------------------- */
+    JPEGXRDataset* poDS = new JPEGXRDataset();
+    JPEGXRDecoder& oDecoder = poDS->oDecoder;
+
+    int nBands = 0;
+
+/* -------------------------------------------------------------------- */
+/*      Initialize decoder and describe source image.                   */
+/* -------------------------------------------------------------------- */
+    try
+    {
+        oDecoder.Initialize(poOpenInfo->pszFilename);
+        oDecoder.GetSize(poDS->nRasterXSize, poDS->nRasterYSize);
+        oDecoder.GetSamplePerPixel(nBands);
+    }
+    catch (JPEGXRException const& e)
+    {
+        CPLError(CE_Failure, CPLE_AppDefined,
+            "JPEGXR - Decoder initialization failed : %s", e.what());
+        return nullptr;
+    }
+
+/* -------------------------------------------------------------------- */
+/*      Create band information objects.                                */
+/* -------------------------------------------------------------------- */
+    CPLAssert(nBands > 0);
+
+    for (int iBand = 1; iBand <= nBands; iBand++)
+    {
+        poDS->SetBand(iBand, new JPEGXRRasterBand(poDS, iBand));
+    }
+
+    CPLAssert(poDS->nBands == nBands);
+
+/* -------------------------------------------------------------------- */
+/*      Initialize any PAM information.                                 */
+/* -------------------------------------------------------------------- */
+    poDS->SetDescription(poOpenInfo->pszFilename);
+    poDS->TryLoadXML();
+
+    return poDS;
+}
+
+/************************************************************************/
+/*                             Uncompress()                             */
+/************************************************************************/
+
+CPLErr JPEGXRDataset::Uncompress()
+{
+    try
+    {
+        if (bHasUncompressed)
+            return CE_None;
+
+        size_t nUncompressedSize = nRasterXSize * nRasterYSize * nBands
+            * GDALGetDataTypeSizeBytes(GetRasterBand(1)->GetRasterDataType());
+        pabyUncompressedData = (GByte*)VSI_MALLOC_VERBOSE(nUncompressedSize);
+        if (!pabyUncompressedData)
+            return CE_Failure;
+
+        oDecoder.Read(pabyUncompressedData, 0, 0, nRasterXSize, nRasterYSize);
```

Comment:
@rouault JPEGXR driver as submitted here is a first working version, simplistic in places and there is plenty of room for improvements, for performance especially. The issue is, jxrlib API is fairly obscure, not well documented and it will take time and trial-error attempts I'm plannig to tackle in future. Unless, that is a blocker...